Partial Assembly Formulation

rcarson3 opened the pull request:
This PR adds a complete implementation of the partial assembly operation for both the RHS of our system (AssembleElementVector) and the matrix operator (AssembleElementGrad) action of our system. The primary benefit of the PA version of the assembly operation is it allows us to run things on the GPU. The initial version of the PA operator when run on the GPU currently takes about the same time to run as the full assembly when not using a preconditioner.

rcarson3 committed:
MFEM_FORALL version
Migrated several for loops and RAJA loops over to MFEM_FORALL

rcarson3 committed:
Missed some MFEM_FORALL's from remote machine

rcarson3 committed:
GPU fixes but QuadFuncVecCoeff and QuadFuncCoeff don't work still

rcarson3 committed:
Output files now work with GPU

rcarson3 committed:
Runtime model options for .toml file now enabled for CUDA, OpenMP, and CPU

rcarson3 committed:
Small changes to try and reduce potential device to host operations

rcarson3 committed:
Fix parallel PA bugs
PA had several bugs within it related to parallel runs. The CUDA version is now running correctly and is now faster than the full assembly CPU version.

rcarson3 committed:
Left things out in last commit
Found a few lines that I left out in the last commit

rcarson3 committed:
Nonlinear PA for RHS / assemble element vec and uncrustify
Added the nonlinear PA opeartion for the assemble element vec and a test case for it. Next, I made a few small changes to keep memory on the device as much as possible. Finally, uncrustify was run on everything.

rcarson3 committed:
Bug fix related to PA formulation
Found a few bugs in the PA formulation that was leading the correct results to drift as the sample is deforming.

nrbarton commented on CMakeLists.txt:
```diff
@@ -17,7 +17,7 @@ endif()
 enable_language(C)
 
-set(CMAKE_CXX_STANDARD 11)
+set(CMAKE_CXX_STANDARD 14)
```
You really need C++14 here?

rcarson3 replied:
I don't believe so. I was just able to test it with C++11 just fine. I think I might have accidentally introduced this into the ExaCMech build system as well. I can't remember why though, but I should fix that there as well...

nrbarton commented on src/mechanics_ecmech.cpp:
```diff
@@ -833,7 +338,286 @@ void ExaCMechModel::ModelSetup(const int nqpts, const int nelems, const int spac
+   init_state_vars(_q_matVars0, histInit);
+}
+
+void KinKMBalDDFCCModel::init_state_vars(mfem::QuadratureFunction *_q_matVars0, std::vector<double> hist_init)
```
I have _not_ managed to grok much of this part of the code, and am really just skimming through. But I am a bit surprised to see details such as the KinKMBalDDFCC name finding their way this far up the call stack. Are there things that this level that are actually specific to the details of the material model being employed?

rcarson3 replied:
So, I have a regular ExaCMech base class that these classes are derived off and for all intents and purposes that's where the meat of things occur. However, I decided it would be easier to have child classes that contain the actual templated material classes within them where I can then correctly instantiate the material models. For example the ```KinKMBalDDFCC``` class really makes use of the already defined ```matModelEvptn_FCC_B``` within ```ExaCMech```. However, I didn't like that name and thought the ```KinKMBalDDFCC``` one would be more descriptive of the actual model being used for this class.

If you'd like we could talk more offline if the above wasn't super clear about what these classes are actually for.

rcarson3 committed:
Bug fix mechanics_integraors.cpp

rcarson3 committed:
Simplify cmake build system
All of the MFEM dependencies should now be properly captured in the FindMFEM.cmake file that was borrowed from the LLNL/Serac repository. I should now put conditionals around whether a dependency was included within MFEM or not.

rcarson3 committed:
Update example options.toml file
Added in the partial assembly and runtime options as examples

rcarson3 committed:
Update readme

rcarson3 committed:
Guard against packages that MFEM hasn't been installed with
Now that we pull in whatever MFEM was built with we need to guard against the case that MFEM wasn't built with Conduit. Also, I've added the ParaView datacollection as a possible output, since it is also a binary format and doesn't require any additional libraries to be included.

rcarson3 committed:
Add new meshing tools for VTK, auto-generated, and conversion from FEpX style meshes

rcarson3 commented:
While messing around with other mesh types, I noticed that mfem's VTK reader could cause some issues due to it setting the boundary elements attributes all to the same value. Therefore, I wrote some meshing scripts that allow a user to convert from some different type of meshes (auto-generated mesh, VTK, or FEpX) into the native MFEMv1.0 mesh format. I need to add additional documentation, but as far as I can tell it's working. I'm basing this from messing around with a 2x2x2 auto-generated hex mesh to MFEMv1.0 to VTK to MFEMv1.0 which returns the same mesh and also the same results when run.

The nice thing about these scripts is it does provide us an easy way to make use of the conformal grain boundary meshes that one can get out from Neper (a polycrystal tessellation and meshing software).

Edit: Here's an example of a simple 100 grain simulation that has conformal grain boundaries run using ExaConstit. It has roughly 200k quadratic tetrahedron elements.
![effective_pl_strain_material](https://user-images.githubusercontent.com/5864850/76382379-ce9bc700-6315-11ea-9068-00a06538d1fb.png)

rcarson3 committed:
Bug fixes related to UMAT and debug builds breaking
UMAT files had several errors in it that would cause it to fail if it was using Device memory. The mechanics_test file was also updated which was causing errors related to the UMAT file. Finally the CMakeList files were updated to no longer lead to breaking debug builds with the newest MFEMv4.1 release.

rcarson3 commented:
I've tested this on some fairly large meshes 8e6 elements on 64 nodes with 384 GPUs and 2560 CPUs on Summit with no problem. I haven't run into any issues with this branch. I'm going to revert the cmake back down to requiring C++11 and fix up some minor warnings while compiling and then merge this PR in.

Also, it looks like ExaCMech handles large deformation of quadratic tets just fine at least out to -26% engineering strain:
![sample_crss](https://user-images.githubusercontent.com/5864850/77461687-2cbdb500-6dc0-11ea-9785-42e8fadac1c5.png)
![sample_material](https://user-images.githubusercontent.com/5864850/77461691-2f200f00-6dc0-11ea-90c2-8ea5aa193b58.png)

rcarson3 committed:
Fix compiler warnings

rcarson3 committed:
Revert CMakeList C++14 requirement to C++11

rcarson3 committed:
Relax scaling factor requirements for NR solver
The scaling factor heuristic could be relaxed from a norm_ratio of 0.1 to 0.5

rcarson3 committed:
Small fixes related to compiler warnings and for provided meshes

rcarson3 committed:
README.md update
